src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +27,116 @@ class DictionarySegment : public BaseSegment {
   /**
    * Creates a Dictionary segment from a given value segment.
    */
-  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment);
+  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment) {
+    auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(base_segment);
+    DebugAssert(value_segment, "The parameter \"base_segment\" should be of type ValueSegment<T>.");
+
+    const auto segment_size = value_segment->size();
+
+    _dictionary = std::make_shared<std::vector<T>>();
+    for (const auto& chunk_value : value_segment->values()) {
+      _insert_in_dictionary(chunk_value);
+    }
```

Comment:
```suggestion
    std::set<T> unique_values;
    for (const auto& chunk_value : value_segment->values()) {
      unique_values.insert(value);
    }
    auto _dictionary = std::make_shared<std::vector<T>>(unique_values);

```
I think using `std::set` would be more convenient here. 

- First, sets are the C++ standard containers, which are used to keep track of unique elements 
- Second, if I'm not mistaken, `insert` has `O(log)` complexity, whereas `upper_bound` can have the complexity of `O(n)` on unsorted vectors
- Third, `std::set` are automatically sorted (even if converting to a `std::vector`)
- Forth, in my opinion, `std::set` makes the code look simpler and more understandable

Reply:
Good suggestion. We also think that using set looks more understandable.
However, at the moment we don't know which implementation performs better.
A benchmark would be needed.

However, the advantage of our implementation is that we work directly with the dictionary vector -> so the implementation is more memory efficient and we don't have to copy (O(n)) the set to the vector.

So for now, we will keep our implementation.

Reply:
We would recommend either `std::set`, `std::unordered_set` or `std::vector` with `std::sort` and `std::unique`.